com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/JarSignatureVerifier.java

```diff
@@ -0,0 +1,175 @@
+import java.net.URL;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateExpiredException;
+import java.security.cert.CertificateNotYetValidException;
+import java.security.cert.X509Certificate;
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Splitter;
+import com.sap.cloud.lm.sl.cf.core.util.FileUtils;
+import com.sap.cloud.lm.sl.cf.process.message.Messages;
+import com.sap.cloud.lm.sl.common.SLException;
+
+@Component
+public class JarSignatureVerifier {
+
+    private static final String META_INF = "META-INF";
+    private static final String CN = "CN";
+    private static final int BUFFER_SIZE = 8 * 1024;
+
+    public void verify(List<X509Certificate> targetCertificates, String certificateCN, URL jarURL) {
+        try {
+            JarFile jarFile = openJarFile(jarURL);
+            List<JarEntry> jarEntries = getJarEntries(jarFile);
+            verifyJarEntries(targetCertificates, jarEntries, certificateCN);
+        } catch (IOException | SecurityException | SLException e) {
+            throw new SLException(MessageFormat.format(Messages.COULD_NOT_VERIFY_ARCHIVE_SIGNATURE, e.getMessage()), e);
+        }
+    }
+
+    private JarFile openJarFile(URL jarFileUrl) throws IOException {
+        JarURLConnection connection = (JarURLConnection) toJarUrl(jarFileUrl).openConnection();
+        return connection.getJarFile();
+    }
+
+    private URL toJarUrl(URL jarUrl) throws MalformedURLException {
+        return new URL("jar:" + jarUrl.toString() + "!/");
+    }
+
+    private List<JarEntry> getJarEntries(JarFile jarFile) throws IOException {
+        List<JarEntry> jarEntries = new ArrayList<>();
+        Enumeration<JarEntry> entries = jarFile.entries();
+        while (entries.hasMoreElements()) {
+            JarEntry jarEntry = readJarEntry(jarFile, entries.nextElement());
+            if (!jarEntry.isDirectory()) {
+                jarEntries.add(jarEntry);
+            }
+        }
+        return jarEntries;
+    }
+
+    private JarEntry readJarEntry(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        FileUtils.validatePath(jarEntry.getName());
+        verifySignature(jarFile, jarEntry);
+        return jarEntry;
+    }
+
+    private void verifySignature(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        try (InputStream inputStream = jarFile.getInputStream(jarEntry)) {
+            while (inputStream.read(buffer, 0, buffer.length) != -1) {
+                // Just read. This will throw a SecurityException if a signature/digest check fails.
```

Comment:
Add another comment saying that `JarEntry.getCertificates()` will return null if the stream isn't read first.